tests/suite/upgrade_test.go

```diff
@@ -0,0 +1,246 @@
+		resultsDir, err = framework.CreateResultsDir("ngf-upgrade", version)
+		Expect(err).ToNot(HaveOccurred())
+
+		filename := filepath.Join(resultsDir, fmt.Sprintf("%s.md", version))
+		resultsFile, err = framework.CreateResultsFile(filename)
+		Expect(err).ToNot(HaveOccurred())
+		Expect(framework.WriteSystemInfoToFile(resultsFile, clusterInfo)).To(Succeed())
+	})
+
+	AfterEach(func() {
+		if skipped {
+			Skip("")
+		}
+
+		Expect(resourceManager.DeleteFromFiles(files, ns.Name)).To(Succeed())
+		Expect(resourceManager.Delete([]client.Object{ns})).To(Succeed())
+		resultsFile.Close()
+	})
+
+	It("upgrades NGF with zero downtime", func() {
+		cfg := framework.InstallationConfig{
+			ReleaseName:          releaseName,
+			Namespace:            ngfNamespace,
+			ChartPath:            localChartPath,
+			NgfImageRepository:   *ngfImageRepository,
+			NginxImageRepository: *nginxImageRepository,
+			ImageTag:             *imageTag,
+			ImagePullPolicy:      *imagePullPolicy,
+			ServiceType:          *serviceType,
+			IsGKEInternalLB:      *isGKEInternalLB,
+		}
+
+		type metricsResults struct {
+			metrics  *vegeta.Metrics
+			testName string
+			scheme   string
+		}
+		metricsCh := make(chan *metricsResults)
+
+		type testCfg struct {
+			desc   string
+			port   string
+			target framework.Target
+		}
+
+		tests := []testCfg{
+			{
+				desc: "Send http /coffee traffic",
+				port: "80",
+				target: framework.Target{
+					Method: "GET",
+					URL:    "http://cafe.example.com/coffee",
+				},
+			},
+			{
+				desc: "Send https /tea traffic",
+				port: "443",
+				target: framework.Target{
+					Method: "GET",
+					URL:    "https://cafe.example.com/tea",
+				},
+			},
+		}
+
+		for _, test := range tests {
+			go func(cfg testCfg) {
+				defer GinkgoRecover()
+
+				results, metrics := framework.RunLoadTest(
+					[]framework.Target{cfg.target},
+					100,
+					60*time.Second,
+					cfg.desc,
+					fmt.Sprintf("%s:%s", address, cfg.port),
+					"cafe.example.com",
+				)
+
+				scheme := strings.Split(cfg.target.URL, "://")[0]
+				metricsRes := metricsResults{
+					metrics:  &metrics,
+					testName: fmt.Sprintf("\n## Test: %s\n\n```text\n", cfg.desc),
+					scheme:   scheme,
+				}
+
+				buf := new(bytes.Buffer)
+				encoder := vegeta.NewCSVEncoder(buf)
+				for _, res := range results {
+					res := res
+					Expect(encoder.Encode(&res)).To(Succeed())
+				}
+
+				csvName := fmt.Sprintf("%s.csv", scheme)
+				filename := filepath.Join(resultsDir, csvName)
+				csvFile, err := framework.CreateResultsFile(filename)
+				Expect(err).ToNot(HaveOccurred())
+
+				_, err = fmt.Fprint(csvFile, buf.String())
+				Expect(err).ToNot(HaveOccurred())
+				csvFile.Close()
+
+				output, err := framework.GeneratePNG(resultsDir, csvName, fmt.Sprintf("%s.png", scheme))
+				Expect(err).ToNot(HaveOccurred(), string(output))
+
+				metricsCh <- &metricsRes
+			}(test)
+		}
+
+		// allow traffic flow to start
+		time.Sleep(2 * time.Second)
```

Comment:
Is there a more deterministic way to do this?

Reply:
Things move so fast when you start the goroutine and then immediately call upgrade, so I just wanted a brief delay to at least allow traffic to flow for a couple seconds before upgrading.

Reply:
Not sure about deterministic, because I don't want to just receive a successful response, I want traffic to be flowing from the vegeta call for a few seconds.